File: 백준/Gold/2493. 탑/탑.java
import java.io.*;
import java.util.*;

public class Main {

    // 타워 정보
    private static class Data {
        private int height;
        private int index;
        Data(int height, int index) {
            this.height = height;
            this.index = index;
        }
    }

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        StringBuilder sb = new StringBuilder();

        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());
        br.close();

        Stack<Data> towerStack = new Stack<>();

        for(int i = 1; i < N + 1; i++) {
            int num = Integer.parseInt(st.nextToken());

            while(!towerStack.isEmpty()) {
                Data towerData = towerStack.peek();
                // 왼쪽 타워의 높이가 현재 타워와 같거나 높으면 index출력하고 stack에 저장
                if(towerData.height >= num) {
                    sb.append(towerData.index).append(" ");
                    towerStack.push(new Data(num, i));
                    break;
                }
                // 낮으면 앞으로 필요없으므로 pop
                else {
                    towerStack.pop();
                }
            }
            if(towerStack.isEmpty()) {
                towerStack.push(new Data(num, i));
                sb.append(0).append(" ");
            }
        }
        bw.write(sb.toString());
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/2493. 탑/탑.kt
import java.util.*
import java.io.*

class Data(val height : Int, val idx : Int)

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {

    val N = readLine().toInt()
    val st = StringTokenizer(readLine())

    val stack = Stack<Data>() // 타워 높이 및 위치 저장 스택
    val bw = BufferedWriter(OutputStreamWriter(System.out))

    for(i in 0 until N) {
        val currentHeight = st.nextToken().toInt()

        // 현재 빌딩 높이가 이전 빌딩 높이보다 높으면 수신 불가능
        // 수신 불가능 시 수신 가능한 높이 나올 때까지 pop
        while(stack.isNotEmpty() && stack.peek().height < currentHeight) {
            stack.pop()
        }
        // empty일 경우 현재 빌딩 높이보다 높은 빌딩 없기 때문에 0
        if(stack.isEmpty()) {
            bw.write("0 ")
        }
        else {
            bw.write("${stack.peek().idx} ")
        }
        stack.push(Data(currentHeight, i + 1))
    }
    close()
    bw.flush()
    bw.close()
}